### PR TITLE
[5.5] Fix logical operations precedence + null default param passing

### PR DESCRIPTION
**Pull request by arthurkushman**

### CODE REVIEW

**Comment by browner12** [comment]
what is the benefit of changing the order of these conditional statements?

**Comment by devcircus** [comment]
Probably should keep the current coding style when making changes like this.

**Comment by arthurkushman** [comment]
The benefit is that the simple expression, if it fails, will leave the execution of the heavy one - it is strict technical rule, no one can ignore things like this. Moreover, there is nothing to do about "coding style", there are just precedence changes.

**Comment by arthurkushman** [comment]
In other words it is much faster to execute equation, then call function + pass arguments via the stack and then do a lot of work in this function (probably O(n) loop ops) before returning the result.

**Comment by browner12** [comment]
There **are** coding style changes. You switched from the second conditional being on a new line to being on the same line.

**Comment by browner12** [comment]
Do you have any metrics on performance differences?

**Comment by laurencei** [comment]
Hi @arthurkushman,

I can see this is your first contribution to the Laravel framework. Thanks and welcome to the community.

Based upon many PR's I've seen in the past - there are a couple of things you might want to consider to improve your chances of Taylor accepting this.

Firstly - you are touching three completetly seperate parts of the framework (`Config`, `Transactions` and `Eloquent`) with the one PR. This makes it difficult in the future to pinpoint issues, so there might be benefit in submitting 3 seperate PRs, each targeting the area you want.

The second issue is you are essentially providing a PR to improve performance, yet you've not provided any performance tests before/after this change. You can look at some [blog posts by Taylor here](https://medium.com/@taylorotwell/benchmarking-laravel-symfony-zend-2c01c2b270f8) about how he's tested performance in the past, and use that as a guide to run your own tests. Once you've done that, if you can include the results that show an improvement, then that will help make Taylor's life easier.

**Comment by taylorotwell** [comment]
I am closing this pull request because it lacks sufficient explanation, tests, or both. It is difficult for us to merge pull requests without these things because the change may introduce breaking changes to the framework.

Feel free to re-submit your change with a **thorough** explanation of the feature and tests - integration tests are preferred over unit tests. Please include it's benefit to end users; the reasons it does not break any existing features; how it makes building web applications easier, etc.

Thanks!

**Comment by arthurkushman** [comment]
@browner12 do you asking me about metrics on performance difference between simple comparison operation and the call to function with call to another 2 and the one `Str::contains` foreach loop wich is O(n)?, to be precise in this line of code ex.:
```php
if ($this->transactions > 1 && $this->causedByDeadlock($e)) {
```
foreach goes from Str::contains with `mb_strpos`:
```php
        foreach ((array) $needles as $needle) {
            if ($needle !== '' && mb_strpos($haystack, $needle) !== false) {
                return true;
            }
        }
```
or in the second case where simple comparison && function call $this->isDirty() within 2 separate loops which is O(n^2) difficulty - first called in hasChanges 
```php
        foreach (Arr::wrap($attributes) as $attribute) {
            if (array_key_exists($attribute, $changes)) {
                return true;
            }
        }
```
and the second in getDirty
```php
        foreach ($this->getAttributes() as $key => $value) {
            if (! $this->originalIsEquivalent($key, $value)) {
                $dirty[$key] = $value;
            }
        }
```
?
If so, if we compare those things above - then this question is quite strange.

Hi @laurencei - this is my 2nd pull request and the 2nd declined/closed, I'm starting to form an opinion that laravel/framework don't actually need any performance improvements on database Model/Eloquent =)

PS was this a **thorough** explanation?

**Comment by browner12** [comment]
I'm sorry. your sentence structure was a little hard to follow.   At the surface, what you're suggesting makes complete sense.  However, it is helpful for us to judge the usefulness of this PR if we can see some actual performance improvement metrics. This is because we have to weigh the improvement vs the possibility this could break existing code.  If this saves us nanoseconds, it's something that might not be worth the possible BCs. If you can show it's saving us seconds, then we can weigh the value differently.

Also, Taylor is making a strong push to make sure all submitted PRs are properly tested. If you can add some integration tests to this change, that also helps give us confidence in this PR.  And because this is a PR to **Eloquent**, we are always overly cautious of it since it is such an important part of the FW.

Don't get discouraged because your PRs were closed. I've had plenty of PRs rejected by Laravel, just like every other contributor in here.

**Comment by arthurkushman** [comment]
Everything is ok, thank u
